src/layer/x86/gridsample_x86.cpp

```diff
@@ -0,0 +1,692 @@
+
+        return std::min(length - 1.0f, std::max(coord, 0.0f));
+    }
+};
+
+template<>
+struct compute_coord<GridSample::Reflection, /*align_corner*/ false>
+{
+#if __AVX__
+    __m256 operator()(__m256 length, __m256 coord)
+    {
+        const __m256 border_x = _mm256_sub_ps(length, *(__m256*)_ps256_1);
+
+        __m256 v0p5fp8 = _mm256_set1_ps(0.5f);
+        coord = _mm256_add_ps(coord, v0p5fp8);
+
+        coord = _mm256_and_ps(coord, *(__m256*)_ps256_inv_sign_mask);
+
+        __m256 reflectx_v = _mm256_and_ps(_mm256_sub_ps(coord, length), *(__m256*)_ps256_inv_sign_mask);
+        coord = _mm256_sub_ps(length, reflectx_v);
+
+        coord = _mm256_sub_ps(coord, v0p5fp8);
+
+        _mm256_sub_ps(coord, v0p5fp8);
+
+        coord = _mm256_min_ps(border_x, _mm256_max_ps(coord, _mm256_setzero_ps()));
+
+        return coord;
+    }
+#endif // __AVX__
+    float operator()(int length, float coord)
+    {
+        coord = fabs(coord + 0.5f);
+        coord = length - fabs(coord - length) - 0.5;
+
+        return std::min(length - 1.0f, std::max(coord, 0.0f));
+    }
+};
+
+#include "gridsample_bilinear_compute_blob.h"
+#include "gridsample_bicubic_compute_blob.h"
+#include "gridsample_nearest_compute_blob.h"
+
+} //namespace GridSample_x86_kernel
+
+int GridSample_x86::forward(const std::vector<Mat>& bottom_blobs, std::vector<Mat>& top_blobs, const Option& opt) const
+{
+    using namespace GridSample_x86_kernel;
+    const Mat& bottom_blob = bottom_blobs[0];
+    const Mat& grid = bottom_blobs[1];
+    Mat& top_blob = top_blobs[0];
+    const int elempack = bottom_blob.elempack;
+
+    int channels = bottom_blob.c;
+    int dims = bottom_blob.dims;
+    size_t elemsize = bottom_blob.elemsize;
+
+    int outw, outh, outd;
+    Mat offset_blob, in_bound_blob, value_blob;
+
+    ncnn::Mat grid_p1;
+    if (grid.elempack != 1)
+    {
+        ncnn::convert_packing(grid, grid_p1, 1, opt);
+    }
+    else
+    {
+        grid_p1 = grid;
+    }
+
+    if (dims == 3)
+    {
+        outw = permute_fusion == 0 ? grid_p1.h : grid_p1.w;
+        outh = permute_fusion == 0 ? grid_p1.c : grid_p1.h;
+
+        top_blob.create(outw, outh, channels, elemsize, elempack, opt.blob_allocator);
+        if (top_blob.empty())
+            return -100;
+
+        if (sample_type == GridSample::Bilinear)
+        {
+            offset_blob.create(outw, outh, 4, elemsize, 1, opt.blob_allocator);
+            in_bound_blob.create(outw, outh, 4, elemsize, 1, opt.blob_allocator);
+            value_blob.create(outw, outh, 2, elemsize, 1, opt.blob_allocator);
+            if (offset_blob.empty() || in_bound_blob.empty() || value_blob.empty())
+                return -100;
+
+            in_bound_blob.fill(-1.0f);
+
+            if (padding_mode == GridSample::Zeros)
+            {
+                if (align_corner == 0)
+                {
+                    gridsample_2d_bilinear_compute_blob<GridSample::Zeros, false> op;
```

Comment:
why template functor class instead of template function ?
Template function is enough for non-hotspot block.

Reply:
function can't partial specialization. QAQ